pkg/build/generator/generator.go

```diff
@@ -0,0 +1,243 @@
+package generator
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+// BuildGenerator is a central place responsible for generating new Build objects
+// from BuildConfigs and other Builds.
+type BuildGenerator struct {
+	Client GeneratorClient
+}
+
+type GeneratorClient interface {
+	GetBuildConfig(ctx kapi.Context, name string) (*buildapi.BuildConfig, error)
+	UpdateBuildConfig(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error
+	GetBuild(ctx kapi.Context, name string) (*buildapi.Build, error)
+	CreateBuild(ctx kapi.Context, build *buildapi.Build) error
+	GetImageRepository(ctx kapi.Context, name string) (*imageapi.ImageRepository, error)
+}
+
+type Client struct {
+	GetBuildConfigFunc     func(ctx kapi.Context, name string) (*buildapi.BuildConfig, error)
+	UpdateBuildConfigFunc  func(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error
+	GetBuildFunc           func(ctx kapi.Context, name string) (*buildapi.Build, error)
+	CreateBuildFunc        func(ctx kapi.Context, build *buildapi.Build) error
+	GetImageRepositoryFunc func(ctx kapi.Context, name string) (*imageapi.ImageRepository, error)
+}
+
+func (c Client) GetBuildConfig(ctx kapi.Context, name string) (*buildapi.BuildConfig, error) {
+	return c.GetBuildConfigFunc(ctx, name)
+}
+
+func (c Client) UpdateBuildConfig(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error {
+	return c.UpdateBuildConfigFunc(ctx, buildConfig)
+}
+
+func (c Client) GetBuild(ctx kapi.Context, name string) (*buildapi.Build, error) {
+	return c.GetBuildFunc(ctx, name)
+}
+
+func (c Client) CreateBuild(ctx kapi.Context, build *buildapi.Build) error {
+	return c.CreateBuildFunc(ctx, build)
+}
+
+func (c Client) GetImageRepository(ctx kapi.Context, name string) (*imageapi.ImageRepository, error) {
+	return c.GetImageRepositoryFunc(ctx, name)
+}
+
+// Generate returns new Build object based on BuildRequest object
+func (g *BuildGenerator) Generate(ctx kapi.Context, request *buildapi.BuildRequest) (*buildapi.Build, error) {
+	var newBuild *buildapi.Build
+	if len(request.BuildConfigName) > 0 {
+		glog.Infof("Generating build for config %s", request.BuildConfigName)
+		bc, err := g.Client.GetBuildConfig(ctx, request.BuildConfigName)
+		if err != nil {
+			return nil, err
+		}
+		newBuild, err = g.generateBuildFromConfig(ctx, bc, request.Revision)
+		if err != nil {
+			return nil, err
+		}
+		err = g.Client.UpdateBuildConfig(ctx, bc)
```

Comment:
Checked it and yes etcd returns `EcodeTestFailed` when trying to update resource which was changed in the meantime (the resourceVersion has changed). For this I had to update `image_change_controller.go` logic to apply changes to newly retrieved BuildConfig object, see `updateConfig` method there. So I'm proposing to leave it as is, if user sees an error when trying to kick a build he'll have to retry it. @smarterclayton what do you think?